doc/developers.rst

```diff
@@ -88,3 +88,116 @@ INI File Option            ```name```                      `name`
+several subdirectories.
+
+``$STATEDIR/clones``
+""""""""""""""""""""
+
+This directory keeps git clones of installed packages
+(``$STATEDIR/clones/package/<name>``), packages sources
+(``$STATEDIR/clones/source/<name>``), and package template repositories
+(``$STATEDIR/clones/template/<name>``).
+
+``zkg`` clones the relevant repositories as needed. It can dynamically re-create
+some of these directories as needed, but interfering in that space is not
+recommended. For example, if you remove a clone of an installed package, the
+installation itself will remain live (via the staging mechanism), but ``zkg``
+will no longer be able to refer to all information about the installed package
+(because anything not explicitly captured about the package in
+``$STATEDIR/manifest.json`` is now gone).
+
+Removal of a package (``zkg remove``) removes its clone in ``$STATEDIR/clones``.
+
+``$STATEDIR/scratch``
+"""""""""""""""""""""
+
+When retrieving information on a package that isn't yet installed, or where
+``zkg`` doesn't want to touch the installed code, ``$STATEDIR/scratch/<name>``
+is a clone of the package's git repo at a version (often a git tag) of
+interest. This clone is shallow for any versions that aren't raw SHA-1
+hashes. The information parsed includes the ``zkg.meta`` as well as git
+branch/tag/commit info.
+
+During package installation, ``zkg`` places backups of user-tweakable files into
+``$STATEDIR/scratch/tmpcfg``. ``zkg`` restores these after package installation
+to preserve the user's edits. During package source aggregation, ``zkg`` places
+temporary versions of ``aggregate.meta`` directly into ``$STATEDIR/scratch``.
+
+Creation or unbundling of a package happens via ``$STATEDIR/scratch/bundle``, to
+compose or retrieve information about the bundle. The directory is deleted and
+re-created at the beginning of those operations:
+
+- During bundling, ``zkg`` copies installed package repos from
+  ``$STATEDIR/clones/<name>`` into ``$STATEDIR/scratch/bundle/<name>``, and
+  creates fresh git clones in the ``bundle`` directory for any packages not
+  currently installed. It creates a ``.tar.gz`` of the whole directory,
+  initially in the bundle directory, and moves it to where the user specified.
+
+- During unbundling, ``zkg`` reads the bundle manifest as well as the git repos
+  of the contained packages, and moves the package repos from the scratch space
+  into ``$STATEDIR/clones/package/<name>``.
+
+When installing a package's ``script_dir`` or ``plugin_dir`` into a staging area
+and the source file is a tarfile, ``zkg`` temporarily extracts the tarball into
+``$STATEDIR/scratch/untar/``.
+
+There's little or no cleanup of files in the scratch space after the operations
+creating them complete. ``zkg`` only deletes, then (re-)creates, the directories
+involved upon next use.
+
+When ``zkg`` isn't in the middle of executing any commands, you can always
+delete the scratch space without negatively affecting ``zkg``.
```

Comment:
I guess that's on your mind, inverting this behavior and introducing `--keep-scratch` / `--keep-testing` or some `--keep-temporary` wildcard would be less surprising.

Reply:
Yeah, especially for `scratch`. `zkg` could no doubt clean up properly in there. And for `testing` it could at least clean up when the tests pass. (And it should be much better about pointing you at the trouble if they don't, but that's another one.)